src/main/java/com/google/step/datamanager/DealManagerDatastore.java

```diff
@@ -99,54 +139,168 @@ public Deal updateDeal(Deal deal) {
-    return new ArrayList<Deal>();
+    Collections.sort(
+        deals,
+        new Comparator<Deal>() {
+          @Override
+          public int compare(Deal deal1, Deal deal2) {
+            return voteManager.getVotes(deal2.id) - voteManager.getVotes(deal1.id); // Descending
+          }
+        });
+    return deals;
   }
 
+  /** Sorts deals based on new (Newest to oldest) */
   @Override
   public List<Deal> sortDealsBasedOnNew(List<Deal> deals) {
-    return new ArrayList<Deal>();
+    Collections.sort(
+        deals,
+        new Comparator<Deal>() {
+          @Override
+          public int compare(Deal deal1, Deal deal2) {
+            return LocalDateTime.parse(deal2.timestamp)
+                .compareTo(LocalDateTime.parse(deal1.timestamp)); // Descending
+          }
+        });
+    return deals;
   }
 
-  @Override
-  public List<Deal> getTrendingDeals() {
-    return new ArrayList<Deal>();
+  /** Sorts deals based on hot score (Highest to lowest) */
+  private List<Deal> sortDealsBasedOnHotScore(List<Deal> deals) {
+    Collections.sort(
+        deals,
+        new Comparator<Deal>() {
+          @Override
+          public int compare(Deal deal1, Deal deal2) {
+            return -Double.compare(deal1.getHotScore(), deal2.getHotScore()); // Descending
+          }
+        });
+    return deals;
+  }
+
+  /** Sorts deals based on hot score (Highest to lowest) */
+  private double epochSeconds(String timestamp) {
+    LocalDateTime time = LocalDateTime.parse(timestamp);
+    long epoch = time.atZone(ZoneId.of(LOCATION)).toEpochSecond();
+    return epoch;
   }
 
   /**
-   * Returns a Deal object transformed from a deal entity.
-   *
-   * @param dealEntity Deal entity.
-   * @return a Deal object transformed from the entity.
+   * Calculates a hot score for each deal entity, which takes into account both the time and the
+   * amount of votes it got
    */
-  private Deal transformEntitytoDeal(Entity dealEntity) {
-    long id = dealEntity.getKey().getId();
-    String description = (String) dealEntity.getProperty("description");
-    String photoBlobkey = (String) dealEntity.getProperty("photoBlobkey");
-    String start = (String) dealEntity.getProperty("start");
-    String end = (String) dealEntity.getProperty("end");
-    String source = (String) dealEntity.getProperty("source");
-    long posterId = (long) dealEntity.getProperty("posterId");
-    long restaurantId = (long) dealEntity.getProperty("restaurantId");
-    return new Deal(id, description, photoBlobkey, start, end, source, posterId, restaurantId);
+  private double calculateHotScore(Entity dealEntity) {
+    int netVotes = voteManager.getVotes(dealEntity.getKey().getId());
+    double order = Math.log(Math.max(Math.abs(netVotes), 1));
+    int sign = 0;
+    if (netVotes > 0) sign = 1;
+    else if (netVotes < 0) sign = -1;
+    double seconds =
+        epochSeconds((String) dealEntity.getProperty("timestamp")) - OLDEST_DEAL_TIMESTAMP;
```

Comment:
what's the point of `OLDEST_DEAL_TIMESTAMP` ? I think removing it won't affect the sorting

Reply:
i just copied from reddit's hot algorithm lol which takes into account the datetime of the oldest post

Reply:
ok